xcdat/regridder/grid.py

```
@@ -532,9 +532,22 @@ def create_grid(
 
         ds = ds.assign_coords({coords.name: coords})
 
+    ds["mask"] = create_mask(ds)
+
     return ds
 
 
+def create_mask(ds: xr.Dataset, dims: Optional[List[CFAxisKey]] = None) -> xr.DataArray:
+    if dims is None:
+        dims = ["X", "Y", "Z"]
+
+    dims = list(dims)
+
+    mask_shape = {ds.cf[x].name: ds.cf[x].shape[0] for x in dims if x in ds.cf.axes}
+
+    return xr.DataArray(np.ones(list(mask_shape.values())), dims=list(mask_shape))
```

Since `create_mask()` is a public function, we should add some direct unit tests (in addition to the one you added).

If I understand correctly, this function returns a `mask` DataArray with same dimensions and shape as the grid and filled with ones (via `np.ones`). Is this function the same as the implementation in the xESMF [example](https://xesmf.readthedocs.io/en/latest/notebooks/Masking.html#Regridding-with-a-mask)? In the xESMF example, they create a `mask` variable based on the values of the `zeta` variable, which can include `np.nan`.

---

@chengzhuzhang Can you review the above code too?

---

@tomvothecoder I can add tests specifically for this. This code is only there to create a default mask that passes through all data, it makes no assumptions on what the user intends to do. It's up the user to modify it afterwards or overwrite like the xESMF example. Internally this more of a convenience thing, as it provides the ability to get the shape of the grid using `ds.mask.shape` which was previously more involved with something like `(ds.lat.shape[0], ds.lon.shape[0])`.

I think a better approach to accomplish the xESMF example would be something along the following:
``` python
ds = xc.open_dataset(...)
ds.regridder.create_mask_from_nans("zeta")
ds["mask"].plot()
```
@chengzhuzhang I think it's really just a preference thing. I went with (0, 1) since xESMF/esmf uses this notation. Not sure if this is applicable in this domain and would depend on the implementation but (0, 1) offers the ability to let values "bleed through" but using fractional values.

---

> This code is only there to create a default mask that passes through all data, it makes no assumptions on what the user intends to do. It's up the user to modify it afterwards or overwrite like the xESMF example. Internally this more of a convenience thing, as it provides the ability to get the shape of the grid using `ds.mask.shape` which was previously more involved with something like `(ds.lat.shape[0], ds.lon.shape[0])`.

Got it, I see what you're saying. It also makes it easier to get the shape of the grid compared to `ds.regridder.grid.shape`. 

If the values of the default mask are `np.ones` and the user does not update the mask, does xESMF ignore the `mask` variable as a no-operation? Just curious if xESMF might interpret the `mask` incorrectly somehow with `np.ones` (when the user doesn't update the values).

### Alternative Solution
> ```python
> ds = xc.open_dataset(...)
> ds.regridder.create_mask_from_nans("zeta")
> ds["mask"].plot()
> ```
> 
> @chengzhuzhang I think it's really just a preference thing. I went with (0, 1) since xESMF/esmf uses this notation. Not sure if this is applicable in this domain and would depend on the implementation but (0, 1) offers the ability to let values "bleed through" but using fractional values.

An alternative solution is to add a flag like `create_mask` where the `mask` variable is created directly from the variable being operated on. The user wouldn't have to worry about creating/updating the `mask` variable downstream and it ensures the correct `mask` is created. 

For example:
```python
ds = xc.open_dataset(...)
ds_regrid = ds.regridder.horizontal("zeta", ...create_mask=True)
ds_regrid["mask"].plot()
```

Any thoughts @chengzhuzhang @xCDAT/core-developers?

---

> Got it, I see what you're saying. It also makes it easier to get the shape of the grid compared to ds.regridder.grid.shape.
>
> If the values of the default mask are np.ones and the user does not update the mask, does xESMF ignore the mask variable as a no-operation? Just curious if xESMF might interpret the mask incorrectly somehow with np.ones (when the user doesn't update the values).

I believe xESM will use it as long as it's there. I don't think there would be any side effect from always having it as ones. I would need to look but if it uses masks like `regrid2` then the math is probably something like `sum(var * mask) / ...` so all ones would let all the values through.

> An alternative solution is to add a flag like create_mask where the mask variable is created directly from the variable being operated on. The user wouldn't have to worry about creating/updating the mask variable downstream and it ensures the correct mask is created.
>
> For example:
> 
> ```
> ds = xc.open_dataset(...)
> ds_regrid = ds.regridder.horizontal("zeta", ...create_mask=True)
> ds_regrid["mask"].plot()
> ```

I think this would be a nice option, I'm between `create_mask` and a little more verbose `create_nan_mask`. If it's common to make a mask from nans we could just enable it by default as well.

---

@jasonb5 Got it. I think we should split out this `"mask"` feature as a another GitHub issue/PR since it is a separate enhancement. We can think more about about how to create the mask variable there too.

That way we can merge this PR focused on outputting regridding weights to release v0.9.0 soon.

---

I just noticed you pushed commits for the `"mask"` feature actually. Since it look almost complete, we can include it in this PR.